Support for modifying source in compare clipboard editor

SougandhS opened the pull request:
For https://github.com/eclipse-platform/eclipse.platform/pull/1862#issuecomment-3028138890

follow-up to https://github.com/eclipse-platform/eclipse.platform/pull/1862

Supports modifying the source from compare editor itself

https://github.com/user-attachments/assets/2f1b60b2-8786-491c-adb1-e30e751009ee

SougandhS commented:
One more thing I have to handle

SougandhS commented:
https://github.com/user-attachments/assets/21b185b3-140b-4ceb-9699-72bfaaadcbc0

SougandhS commented on team/bundles/org.eclipse.compare/compare/org/eclipse/compare/internal/ClipboardCompare.java:
```diff
@@ -74,32 +89,38 @@ protected boolean isEnabled(ISelection selection) {
-			editor = fromEditor.getActiveEditor();
+		if (editor instanceof MultiPageEditorPart mpe) {
+			Object page2 = mpe.getSelectedPage();
+			if (page2 instanceof EditorPart e) {
```
Done

SougandhS committed:
Support Source Edit in Compare with Clipboard

Support for modifying source in compare clipboard editor and Use
MultiPageEditorPart for getting active editor and in ClipboardReplace &
Clipboard Compare

SougandhS commented:
> @SougandhS can you solve the merge conflict?

Done 👍

vogella reviewed (commented):
A few review notes (haven't pulled the branch, just read the diff):

1. Likely bug in `setContent` (around line 218): the branch `if (selectionLength <= 1)` overwrites the entire file with `newContent`. With a 1‑character partial selection, this silently replaces the whole file with just the edited char. The condition should be `if (!partialSelection)`, mirroring how `getContents` decides.

2. Mutable per-action state on a singleton delegate: `currentResouce`, `offSet`, `len`, `partialSelection` are now instance fields mutated inside the per-file loop and later read by `EditableFileNode#setContent` after the compare editor is open. A later invocation can flip behavior of an already-open compare editor. Pass these as parameters / capture them in the node instead.

3. Sentinel mixed with `Integer.MAX_VALUE`: `offSet`/`len` use `-1` as "no selection", and the else branch then constructs `new EditableFileNode(currentResouce, 0, Integer.MAX_VALUE)`. It only works because `getContents` clamps via `Math.min`. A single boolean inside the node would be cleaner.

4. Dialogs from model code: `EditableFileNode#getContents` and `setContent` call `MessageDialog.openError(...)` directly; these can be invoked off the UI thread by the compare framework. Prefer `ILog.of(getClass()).error(...)` and rethrow / return, so the framework surfaces the error.

5. Cosmetics: typo `currentResouce` (should be `currentResource`), field `offSet` (should be `offset`), variable name `page2`, and `ClipboardCompare.java` is missing a trailing newline. Pre-existing typo `"Comparision Failed"` could be fixed in passing.

6. No tests added; given the offset/length math, a test that opens a clipboard compare on a partial selection and writes back would be valuable.

7. Edge case: captured `selectionOffset`/`selectionLength` become stale if the underlying file is modified between opening the compare and saving.

SougandhS commented:
>A few review notes (haven't pulled the branch, just read the diff):

Thanks, will check this comments 👍

SougandhS commented:
> A few review notes (haven't pulled the branch, just read the diff):
> 
> 1. Likely bug in `setContent` (around line 218): the branch `if (selectionLength <= 1)` overwrites the entire file with `newContent`. With a 1‑character partial selection, this silently replaces the whole file with just the edited char. The condition should be `if (!partialSelection)`, mirroring how `getContents` decides.

Fixed 👍 

> 2. Mutable per-action state on a singleton delegate: `currentResouce`, `offSet`, `len`, `partialSelection` are now instance fields mutated inside the per-file loop and later read by `EditableFileNode#setContent` after the compare editor is open. A later invocation can flip behavior of an already-open compare editor. Pass these as parameters / capture them in the node instead.

Fixed 👍 

> 3. Sentinel mixed with `Integer.MAX_VALUE`: `offSet`/`len` use `-1` as "no selection", and the else branch then constructs `new EditableFileNode(currentResouce, 0, Integer.MAX_VALUE)`. It only works because `getContents` clamps via `Math.min`. A single boolean inside the node would be cleaner.

Refactored 👍 

> 4. Dialogs from model code: `EditableFileNode#getContents` and `setContent` call `MessageDialog.openError(...)` directly; these can be invoked off the UI thread by the compare framework. Prefer `ILog.of(getClass()).error(...)` and rethrow / return, so the framework surfaces the error.

Used `ILog.` 👍 

> 5. Cosmetics: typo `currentResouce` (should be `currentResource`), field `offSet` (should be `offset`), variable name `page2`, and `ClipboardCompare.java` is missing a trailing newline. Pre-existing typo `"Comparision Failed"` could be fixed in passing.

Refactored 👍 

> 6. No tests added; given the offset/length math, a test that opens a clipboard compare on a partial selection and writes back would be valuable.

handled this now in a simpler way now 👍 

> 7. Edge case: captured `selectionOffset`/`selectionLength` become stale if the underlying file is modified between opening the compare and saving.

Fixed this huge bug 👍
